src/main/java/com/commercetools/sync/commons/utils/GenericUpdateActionUtils.java

```diff
@@ -0,0 +1,194 @@
+    }
+
+    // This method is not private since it is used by one of the unit tests.
+    @Nonnull
+    static <T extends Custom & Resource<T>> Optional<UpdateAction<T>> buildTypedUpdateAction(
+            @Nonnull final T resource,
+            @Nonnull final String updateActionName) throws BuildUpdateActionException {
+        return buildTypedUpdateAction(null, null, null, null,
+                resource, updateActionName);
+    }
+
+    @Nonnull
+    private static <T extends Custom & Resource<T>> Optional<UpdateAction<T>> buildTypedUpdateAction(
+            @Nullable final String customFieldName,
+            @Nullable final JsonNode customFieldValue,
+            @Nonnull final T resource,
+            @Nonnull final String updateActionName) throws BuildUpdateActionException {
+        return buildTypedUpdateAction(null, null, customFieldName, customFieldValue,
+                resource, updateActionName);
+
+    }
+
+    /**
+     * Creates a CTP update action on the given resource {@link T} (which currently could either be a {@link Category}
+     * or a {@link Channel}) according to the {@code updateActionName} flag. According to this flag value, the required
+     * update action is built:
+     * <ol>
+     * <li>SET_CUSTOM_TYPE_REMOVE -> creates a "setCustomType" update action that removes a custom type from the resource.</li>
+     * <li>SET_CUSTOM_TYPE -> creates a "setCustomType" update action that changes the custom type set on the resource.</li>
+     * <li>SET_CUSTOM_FIELD -> creates a "setCustomField" update action that changes the value of a custom field.</li>
+     * </ol>
+     *
+     * @param customTypeKey       the key of the new custom type, only if the flag is SET_CUSTOM_TYPE.
+     * @param customFieldsJsonMap the custom fields map of JSON values, only if the flag is SET_CUSTOM_TYPE.
+     * @param customFieldName     the name of the custom field to update, only if the flag is SET_CUSTOM_FIELD.
+     * @param customFieldValue    the new JSON value of the custom field, only if the flag is SET_CUSTOM_FIELD.
+     * @param resource            the resource to do the update action on.
+     * @param updateActionName    the flag value that decided which update action to do.
+     * @param <T>                 the type of the resource to do the update action on.
+     * @return an update action that depends on the provided flag on the resource it's requested on.
+     */
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    private static <T extends Custom & Resource<T>> Optional<UpdateAction<T>> buildTypedUpdateAction(
+            @Nullable final String customTypeKey,
+            @Nullable final Map<String, JsonNode> customFieldsJsonMap,
+            @Nullable final String customFieldName,
+            @Nullable final JsonNode customFieldValue,
+            @Nonnull final T resource,
+            @Nonnull final String updateActionName) throws BuildUpdateActionException {
+        if (resource instanceof Category) {
```

Comment:
Thanks for the idea. This would make it easier and cleaner to extend, also more robust.
a771980